Added Delete Index support to high-level REST client

catalin-ursachi opened the pull request:
Added high-level REST client support for one of the endpoints covered by https://github.com/elastic/elasticsearch/issues/25847. Will try and add some more, if this one's fine. 🙂 

(Haven't been able to get `gradle check` to run through yet - due to error:
```
:docs:integTestCluster#installAnalysisIcuPlugin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':docs:integTestCluster#installAnalysisIcuPlugin'.
> A problem occurred starting process 'command 'cmd''
```
but ES builds fine, and the newly added tests pass.)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

s1monw commented:
@elasticmachine ok to test

s1monw commented:
@catalin-ursachi this is awesome. It's exactly how we envisioned this client to grow. We will go through the code with you soon but bear with us we have a large team meeting this week so it might be delayed a bit. Thanks for opening this PR

s1monw commented:
@catalin-ursachi FYI you can just `cd client/rest-high-level/` and run`gradle check` from there no need to run all tests here on you end. Also `gradle precommit` is useful as it allows you to only run stuff like checkstyle etc.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java:
```diff
@@ -378,6 +390,11 @@ static String endpoint(String[] indices, String[] types, String endpoint) {
+
     /**
      * Utility method to build request's endpoint.
      */
```
Can we add a unit test method to RequestTests for the new deleteIndex method?

javanna commented on core/src/main/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexResponse.java:
```diff
@@ -48,4 +51,34 @@ public void writeTo(StreamOutput out) throws IOException {
         writeAcknowledged(out);
     }
+
+    public static DeleteIndexResponse fromXContent(XContentParser parser) throws IOException {
```
We need to add unit tests for this parsing method, similar to what we did in `IndexResponseTests`, `GetResponseTests` etc.

javanna commented on core/src/main/java/org/elasticsearch/action/support/master/AcknowledgedResponse.java:
```diff
@@ -61,4 +68,41 @@ protected void readAcknowledged(StreamInput in) throws IOException {
+    }
+
+    protected static void parseInnerToXContent(XContentParser parser, AcknowledgedResponse.Builder context) throws IOException {
+        XContentParser.Token token = parser.currentToken();
```
given this hierarchy, maybe we should check already if we support classes that add their custom fields to the response, like `CreateIndexResponse`. Maybe we can add parsing code for that one too (only parsing, without adding the rest of the API)? What do you think?

catalin-ursachi commented:
Thanks for the review; I'll write the changes as soon as I get the time (possibly Sunday). 🙂

catalin-ursachi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java:
```diff
@@ -378,6 +390,11 @@ static String endpoint(String[] indices, String[] types, String endpoint) {
     }
 
+
+    static String endpoint(String[] indices, String endpoint) {
```
The problem is that it would then clash with the method below:

```
    /**
     * Utility method to build request's endpoint.
     */
    static String endpoint(String... parts) {
        StringJoiner joiner = new StringJoiner("/", "/", "");
        for (String part : parts) {
            if (Strings.hasLength(part)) {
                joiner.add(part);
            }
        }
        return joiner.toString();
    }
```

To be fair, having multiple `endpoint` methods that all take String collections as parameters, but have different behaviour, does feel like asking for trouble. How about renaming one or the other to `buildEndpoint`, `buildPath`, or similar?

catalin-ursachi replied:
Alternatively, I could just not add a new method and call:

```
  String endpoint = endpoint(deleteIndexRequest.indices(), new String[0], "");
```

Would that be preferable?

catalin-ursachi commented on core/src/main/java/org/elasticsearch/action/support/master/AcknowledgedResponse.java:
```diff
@@ -61,4 +68,41 @@ protected void readAcknowledged(StreamInput in) throws IOException {
     }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
Ah, sorry, that was just some over-enthusiastic copy & pasting.

javanna replied:
no worries!

catalin-ursachi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java:
```diff
@@ -378,6 +390,10 @@ static String endpoint(String[] indices, String[] types, String endpoint) {
         return endpoint(String.join(",", indices), String.join(",", types), endpoint);
     }
 
+    static String endpoint(String[] indices, String endpoint) {
```
(see comment thread: https://github.com/elastic/elasticsearch/pull/27019#discussion_r145629135)

javanna replied:
thanks for your comments, how about using the existing endpoint method then and calling it like this `endpoint(indices, Strings.EMPTY_ARRAY, "")` ?

catalin-ursachi commented:
Addressed all the review comments (still unsure about https://github.com/elastic/elasticsearch/pull/27019#discussion_r145629135, however). Will write the docs shortly.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/5.6/indices.html">Indices API on elastic.co</a>
+ */
+public class IndicesClient {
```
should we make this final?

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -221,6 +221,15 @@ public final void close() throws IOException {
     /**
+     * Provides an {@link IndicesClient} which can be used to access the Indices API.
+     *
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/5.6/indices.html">Indices API on elastic.co</a>
```
can you replace the version number with current like in the other links in this class?

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * A wrapper for the {@link RestHighLevelClient} that provides methods for accessing the Indices API.
+ *
+ * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/5.6/indices.html">Indices API on elastic.co</a>
```
can you replace 5.6 with current?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private void createIndex(String index) throws IOException {
+        Response response = client().performRequest("PUT", "/" + index);
```
I guess you don't need the "/" in front of the index name here. the client will do it for you if I remember correctly

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java:
```diff
@@ -0,0 +1,68 @@
+        assertEquals(200, response.getStatusLine().getStatusCode());
+    }
+
+    private boolean indexExists(String index) throws IOException {
```
can we make this static?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    private void createIndex(String index) throws IOException {
```
can this be static?

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    /**
+     * Asynchronously deletes an index using the Delete Index api
```
can you make api capital in all of its occurrences please?

catalin-ursachi commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java:
```diff
@@ -240,6 +243,44 @@ private static void getAndExistsTest(Function<GetRequest, Request> requestConver
+            deleteIndexRequest.indicesOptions(IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(),
+                randomBoolean()));
+        }
+        expectedParams.put("ignore_unavailable", Boolean.toString(deleteIndexRequest.indicesOptions().ignoreUnavailable()));
```
That's what I intended; however, the `DeleteIndexRequest` and the `SearchRequest` don't share any superclass; they define the `indicesOptions` methods separately. So a common class would have to take getter and setter lambdas as arguments; would that be preferable?

javanna replied:
I like the lambdas idea, maybe you can do the same with `timeout`.

catalin-ursachi commented on core/src/main/java/org/elasticsearch/action/support/master/AcknowledgedResponse.java:
```diff
@@ -61,4 +66,10 @@ protected void readAcknowledged(StreamInput in) throws IOException {
         out.writeBoolean(acknowledged);
     }
+
+    protected static final ParseField ACKNOWLEDGED = new ParseField("acknowledged");
```
Indeed it doesn't.

catalin-ursachi commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexResponseTests.java:
```diff
@@ -62,4 +70,74 @@ public void testSerializationWithOldVersion() throws IOException {
+        builder.field("acknowledged", acknowledged);
+        builder.field("shards_acknowledged", shardsAcked);
+        builder.field("index", index);
+        builder.endObject();
```
I'd be tempted to add a TODO here, asking for the test to be updated once  `AcknowledgedResponse's implement `ToXContent`. Would that be okay, or would you rather I changed 'Create Index' and 'Delete Index' to implement `ToXContent` now?

catalin-ursachi replied:
Never mind the above; I've modified `CreateIndexResponse` and `DeleteIndexResponse` to implement `ToXContentObject`, and added a TODO in `AcknowledgedRestListener.java`.

javanna replied:
++ thanks I think there is some value in this from a testing point of view, thank you!

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java:
```diff
@@ -281,6 +269,8 @@ public void testDeleteIndex() throws IOException {
     }
 
+
+
```
nit: may I ask you to remove these two empty lines?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java:
```diff
@@ -944,13 +921,33 @@ private static void randomizeFetchSourceContextParams(Consumer<FetchSourceContex
+            expectedParams.put("expand_wildcards", "closed");
+        } else {
+            expectedParams.put("expand_wildcards", "none");
+        }
```
I wonder if we should simplify this method. Given that we generate the random indices options as part of the method, we could reuse those rather than using the getter, then the getter argument is not needed. Also you could just return the indices options and that would remove the need for the setter argument too. Thoughts?

catalin-ursachi replied:
But if we don't generate an `IndicesOptions`, the `getter` then gets the default for the given request. Without that, we'd have to pass in the default to this method.

So, we could change it to:

```
private static IndicesOptions createAndExpectRandomIndicesOptions(IndicesOptions default, Map<String, String> expectedParams) {
  if (randomBoolean()) {
    ...
  } else {
    //use default
    return null;
  }
```

And then handle the returned value depending on whether it's null or not. Is that nicer, though?

javanna replied:
good point! I missed the default case. I do not like the `null` invariant that we would introduce here, leave it as-is. Thanks for your patience ;)

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -365,7 +376,7 @@ public void searchAsync(SearchRequest searchRequest, ActionListener<SearchRespon
 
     /**
-     * Executes a search using the Search Scroll api
+     * Executes a search using the Search Scroll API
```
good catch :) thanks for adjusting these

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java:
```diff
@@ -240,6 +245,32 @@ private static void getAndExistsTest(Function<GetRequest, Request> requestConver
+    }
+
+
+
```
may I ask you to remove these empty lines and leave just one?

javanna commented on docs/java-rest/high-level/apis/deleteindex.asciidoc:
```diff
@@ -0,0 +1,75 @@
+include-tagged::{doc-tests}/IndicesClientDocumentationIT.java[delete-index-request-timeout]
+--------------------------------------------------
+<1> Timeout to wait for primary shard to become available as a `TimeValue`
+<2> Timeout to wait for primary shard to become available as a `String`
```
This is rather the "Timeout to wait for the all the nodes to acknowledge the index deletion"

javanna commented on docs/java-rest/high-level/apis/deleteindex.asciidoc:
```diff
@@ -0,0 +1,75 @@
+<2> Called in case of failure. The raised exception is provided as an argument
+
+[[java-rest-high-delete-index-response]]
+==== Delete Response
```
make it "Delete Index Response" ?

javanna commented on docs/java-rest/high-level/apis/deleteindex.asciidoc:
```diff
@@ -0,0 +1,75 @@
+[[java-rest-high-delete-index-response]]
+==== Delete Response
+
+The returned `DeleteResponse` allows to retrieve information about the executed
```
`DeleteIndexResponse` rather than `DeleteResponse`

javanna commented on docs/java-rest/high-level/apis/deleteindex.asciidoc:
```diff
@@ -0,0 +1,75 @@
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/IndicesClientDocumentationIT.java[delete-index-response]
```
could you add a little bullet point explaining what the `acknowledged` field holds? "Indicates whether all of the nodes have acknowledged the request or not"

javanna commented:
retest this please

javanna commented on docs/java-rest/high-level/apis/index.asciidoc:
```diff
@@ -1,4 +1,5 @@
 include::_index.asciidoc[]
+include::deleteindex.asciidoc[]
```
can we move this one line above?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java:
```diff
@@ -903,13 +919,43 @@ private static void randomizeFetchSourceContextParams(Consumer<FetchSourceContex
+        }
+    }
+
+    private static <T> void setRandomTimeout(Function<String, T> setter, TimeValue defaultTimeout, Map<String, String> expectedParams) {
```
here and in setRandomIndicesOptions, you could replace the `Function` with a `Consumer`. That works and makes the generics go away.

catalin-ursachi commented:
The failure looks unrelated to the changes; could you please retest?

tvernum commented:
@elasticmachine test this please

javanna commented:
retest this please

catalin-ursachi commented:
No idea if merging in master fixed this (had it become incompatible with some external dependency?), or it was just a series of flukes; but, anyway.

javanna commented:
@catalin-ursachi I looked at the failures too today and I am not sure what happened there, something broke badly but it was not certainly your fault, merging fixes it, happy that the build is green now, I will merge this in tomorrow.

javanna commented:
Thanks a lot @catalin-ursachi , great job!
